src/boards/include/boards/pimoroni_motor2040.h

```diff
@@ -0,0 +1,246 @@
+#endif
+
+#ifndef MOTOR2040_ENCODER_A_B_PIN
+#define MOTOR2040_ENCODER_A_B_PIN 1
+#endif
+
+#ifndef MOTOR2040_ENCODER_B_A_PIN
+#define MOTOR2040_ENCODER_B_A_PIN 2
+#endif
+
+#ifndef MOTOR2040_ENCODER_B_B_PIN
+#define MOTOR2040_ENCODER_B_B_PIN 3
+#endif
+
+#ifndef MOTOR2040_ENCODER_C_A_PIN
+#define MOTOR2040_ENCODER_C_A_PIN 12
+#endif
+
+#ifndef MOTOR2040_ENCODER_C_B_PIN
+#define MOTOR2040_ENCODER_C_B_PIN 13
+#endif
+
+#ifndef MOTOR2040_ENCODER_D_A_PIN
+#define MOTOR2040_ENCODER_D_A_PIN 14
+#endif
+
+#ifndef MOTOR2040_ENCODER_D_B_PIN
+#define MOTOR2040_ENCODER_D_B_PIN 15
+#endif
+
+#ifndef MOTOR2040_NUM_ENCODERS
+#define MOTOR2040_NUM_ENCODERS 4
+#endif
+
+#ifndef MOTOR2040_TX_TRIG_PIN
+#define MOTOR2040_TX_TRIG_PIN 16
+#endif
+
+#ifndef MOTOR2040_RX_ECHO_PIN
+#define MOTOR2040_RX_ECHO_PIN 17
+#endif
+
+#ifndef MOTOR2040_LED_DATA_PIN
+#define MOTOR2040_LED_DATA_PIN 18
+#endif
+
+#ifndef MOTOR2040_NUM_LEDS
+#define MOTOR2040_NUM_LEDS 1
+#endif
+
+#ifndef MOTOR2040_I2C
+#define MOTOR2040_I2C 0
+#endif
+
+#ifndef MOTOR2040_INT_PIN
+#define MOTOR2040_INT_PIN 19
+#endif
+
+#ifndef MOTOR2040_SDA_PIN
+#define MOTOR2040_SDA_PIN 20
+#endif
+
+#ifndef MOTOR2040_SCL_PIN
+#define MOTOR2040_SCL_PIN 21
+#endif
+
+#ifndef MOTOR2040_USER_SW_PIN
+#define MOTOR2040_USER_SW_PIN 23
+#endif
+
+#ifndef MOTOR2040_ADC_ADDR_0_PIN
+#define MOTOR2040_ADC_ADDR_0_PIN 22
+#endif
+
+#ifndef MOTOR2040_ADC_ADDR_1_PIN
+#define MOTOR2040_ADC_ADDR_1_PIN 24
+#endif
+
+#ifndef MOTOR2040_ADC_ADDR_2_PIN
+#define MOTOR2040_ADC_ADDR_2_PIN 25
+#endif
+
+#ifndef MOTOR2040_ADC0_PIN
+#define MOTOR2040_ADC0_PIN 26
+#endif
+
+#ifndef MOTOR2040_ADC1_PIN
+#define MOTOR2040_ADC1_PIN 27
+#endif
+
+#ifndef MOTOR2040_ADC2_PIN
+#define MOTOR2040_ADC2_PIN 28
+#endif
+
+#ifndef MOTOR2040_NUM_ADC_PINS
+#define MOTOR2040_NUM_ADC_PINS 3
+#endif
+
+#ifndef MOTOR2040_SHARED_ADC_PIN
+#define MOTOR2040_SHARED_ADC_PIN 29
+#endif
+
+#ifndef MOTOR2040_CURRENT_SENSE_A_ADDR
+#define MOTOR2040_CURRENT_SENSE_A_ADDR 0b000
+#endif
+
+#ifndef MOTOR2040_CURRENT_SENSE_B_ADDR
+#define MOTOR2040_CURRENT_SENSE_B_ADDR 0b001
+#endif
+
+#ifndef MOTOR2040_CURRENT_SENSE_C_ADDR
+#define MOTOR2040_CURRENT_SENSE_C_ADDR 0b010
+#endif
+
+#ifndef MOTOR2040_CURRENT_SENSE_D_ADDR
+#define MOTOR2040_CURRENT_SENSE_D_ADDR 0b011
+#endif
+
+#ifndef MOTOR2040_VOLTAGE_SENSE_ADDR
+#define MOTOR2040_VOLTAGE_SENSE_ADDR 0b100
+#endif
+
+#ifndef MOTOR2040_FAULT_SENSE_ADDR
+#define MOTOR2040_FAULT_SENSE_ADDR 0b101
+#endif
+
+#ifndef MOTOR2040_SENSOR_1_ADDR
+#define MOTOR2040_SENSOR_1_ADDR 0b110
+#endif
+
+#ifndef MOTOR2040_SENSOR_2_ADDR
+#define MOTOR2040_SENSOR_2_ADDR 0b111
+#endif
+
+#ifndef MOTOR2040_NUM_SENSORS
+#define MOTOR2040_NUM_SENSORS 2
+#endif
+
+// --- UART ---
+#ifndef PICO_DEFAULT_UART
+#define PICO_DEFAULT_UART 0
+#endif
+#ifndef PICO_DEFAULT_UART_TX_PIN
+#define PICO_DEFAULT_UART_TX_PIN MOTOR2040_TX_TRIG_PIN
+#endif
+#ifndef PICO_DEFAULT_UART_RX_PIN
+#define PICO_DEFAULT_UART_RX_PIN MOTOR2040_RX_ECHO_PIN
+#endif
```

Comment:
They're intended to be dual function, for use with both uart devices and HC-SR04 ultrasounds. The full pinout on the board specifically aligns with the ultrasound's pinout for instance.

Reply:
Given that any code using those pins obviously has to be compiled to either use them as UART pins, **or** as distance-sensor pins, I wonder if it _might_ make sense to do something like:  (extra blank lines added only for clarity)
```
#ifndef MOTOR2040_USE_DISTANCE_SENSOR
// default to 0, unless explicitly set to 1
#define MOTOR2040_USE_DISTANCE_SENSOR 0
#endif

#if MOTOR2040_USE_DISTANCE_SENSOR

#define MOTOR2040_TRIG_PIN 16
#define MOTOR2040_ECHO_PIN 17
// no PICO_DEFAULT_UART

#else

#define MOTOR2040_TX_PIN 16
#define MOTOR2040_RX_PIN 17
// --- UART ---
#ifndef PICO_DEFAULT_UART
#define PICO_DEFAULT_UART 0
#endif
#ifndef PICO_DEFAULT_UART_TX_PIN
#define PICO_DEFAULT_UART_TX_PIN MOTOR2040_TX_PIN
#endif
#ifndef PICO_DEFAULT_UART_RX_PIN
#define PICO_DEFAULT_UART_RX_PIN MOTOR2040_RX_PIN
#endif

#endif
```
:thinking: 

But this is only a suggestion, and obviously if @kilograham has any opinions his view supersedes mine :slightly_smiling_face:

Reply:
that may well be better, but it is up to how you guys expect the board to be used (this is good because you don't want UART going there if you are using it as ultrasound).

Reply:
I like this solution